File: day2/circlepassing/submissions/accepted/wendy.cpp
#include <bits/stdc++.h>
#define f(i, s, k, l) for (int i = s; i < k; i += l)
#define for0(i, k) f(i, 0, k, 1)

#define pl pair<long long, long long>
#define pb push_back

#define vl vector<long long>
#define vi vector<int>
#define sz(x) (ll)(x).size()

using namespace std;
using ll = long long;
using ld = long double;

ll minDirectPass(ll n, ll x, ll y) {
    return min(abs(x - y), n - abs(x - y));
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
    
    ll n, m, q, a, x, y; 
    cin >> n >> m >> q; 
    set<ll> s; 
    for0(i, m) {
        cin >> a; 
        s.insert(a); 
        s.insert(a + n); 
    } 
    n *= 2; 

    for0(i, q) {
        cin >> x >> y; 
        
        // pass directly
        ll res = minDirectPass(n, x, y); 
 
        // right bridge
        auto b = s.upper_bound(x);
        if (b == s.end()) b = s.begin(); 
        res = min(res, minDirectPass(n, x, *b) + minDirectPass(n, (*b + n / 2) % n, y) + 1);   
        // cout << *b << " " << res << endl; 

        // left bridge
        if (b != s.begin()) b = prev(b);
        else b = prev(s.end());  
        res = min(res, minDirectPass(n, x, *b) + minDirectPass(n, (*b + n / 2) % n, y) + 1);   

        cout << res << endl; 
    }
}

File: day2/circlepassing/submissions/accepted/wendy.py
#!/usr/bin/python3
def minDirectPass(n, x, y):
    return min(abs(x - y), n - abs(x - y))

n, m, q = [int(x) for x in input().split()]
s = [int(x) for x in input().split()]
s += [a + n for a in s]

n *= 2

for _ in range(q):
    x, y = [int(x) for x in input().split()]
    res = minDirectPass(n, x, y)
    
    l = 0
    r = len(s) if s[0] < x else 0
    while r - l > 1: 
        mid = (l + r) // 2
        if s[mid] > x: 
            r = mid
        else:
            l = mid 

    # right bridge
    r = 0 if r == len(s) else r
    res = min(res, minDirectPass(n, x, s[r]) + minDirectPass(n, (s[r] + n // 2) % n, y) + 1) 

    # left bridge
    r = r - 1 if r != 0 else len(s) - 1
    res = min(res, minDirectPass(n, x, s[r]) + minDirectPass(n, (s[r] + n // 2) % n, y) + 1)

    print(res)